### ISSUE TITLE
[QUESTION]: How to make prediction with calibrated uncertainties when test_path is set to /dev/null

### ISSUE THREAD

**Issue by schinto** [chemprop/chemprop#324]
Dear ChemProp developers,
1. I would like to load a model once during the startup of a webservice and make predictions with calibrated uncertainties
based on lists of SMILES strings. From the documentation it's not clear how to do the uncertainty calibration when `--test_path` is set to `/dev/null` and `smiles` is used as argument in `chemprop.train.make_predictions`
2. How to silence tqdm and warnings during prediction, so that server logs are not polluted?
3. Is cross-validation done during calibration as in sklearn.calibration.CalibratedClassifierCV ?
4. Is separate calibration data required or can the test be used for calibration of the model?  

Here is my example code. The TODOs illustrate my questions.
````python
from fastapi import FastAPI, Path, HTTPException, Form
from fastapi.responses import Response, RedirectResponse, PlainTextResponse, HTMLResponse
import pathlib
import itertools
import chemprop
from time import perf_counter
from tqdm import tqdm
from functools import partialmethod

# TODO: Try to silence tqdm progress bars
tqdm.__init__ = partialmethod(tqdm.__init__, disable=True)

# Base directory of app
BASE_DIR = pathlib.Path(__file__).resolve(strict=True).parent

app = FastAPI(title="Property Prediction API",
    description="API to predict molecular properties",
    version="0.1"
)

# Global variables for model loading and chemprop arguments
model = None
model_args = None

@app.on_event("startup")
def load_model():
    # TODO: How to silence the model loading? like: Loading pretrained parameter ...
    global model
    global model_args
    arguments = [
        '--test_path', '/dev/null',
        '--preds_path', '/dev/null',
        '--checkpoint_dir', BASE_DIR.joinpath("model_checkpoint").as_posix(),
        '--features_generator', 'rdkit_2d_normalized',
        '--no_features_scaling',
        '--uncertainty_method', 'ensemble',
# TODO: How to get calibrated results ?
# If the following lines are uncommented, then an error is thrown that the test_path is missing
        # '--calibration_method', 'zelikman_interval',
        # '--calibration_path', 'data/calibration_data.csv',
        # '--regression_calibrator_metric', 'stdev'
    ]
    model_args = chemprop.args.PredictArgs().parse_args(arguments)
    model = chemprop.train.load_model(args=model_args)

@app.get("/", include_in_schema=False)
async def docs_redirect():
    return RedirectResponse(url="docs")

def flatten(list2d):
    """Flatten a list of lists to a list"""
    return list(itertools.chain.from_iterable(list2d))

# async run_prediction(smiles):
def run_prediction(smiles):
    """Run prediction"""
    # TODO: How to silence ouput of make_predictions?
    # How to suppress WARNING:root:No normalization for BCUT2D_... ?

    # Create list of lists
    smiles = [ [s] for s in smiles ]
    prediction, uncertainty = chemprop.train.make_predictions(
        model_args, smiles=smiles, model_objects=model,
        return_invalid_smiles=True, return_uncertainty=True,
    )
    # Flatten list of lists
    smiles = flatten(smiles)
    prediction = flatten(prediction)
    uncertainty = flatten(uncertainty)
    return smiles, prediction, uncertainty

@app.get("/predict/{smiles:path}", tags=["predictions"])
async def predict(
    smiles: str = Path(..., title="SMILES for property prediction"),
):
    """Predict property

    Example SMILES: `N#Cc1ccccc1C/C=C\C`
    """
    try:
        smiles_list = [ smiles.strip() ]     
        start_time = perf_counter()
        smiles, prediction, uncertainty = run_prediction(smiles_list)
        # smiles, prediction, uncertainty = await run_prediction(smiles_list)
        end_time = perf_counter()
        execution_time = end_time - start_time
        result = {
            "execution_time": execution_time,
            "data": { "smiles": smiles, "pred": prediction, "uncertainty": uncertainty }
         }
        return result
    except:
        raise HTTPException(
            status_code=404,
            detail=f"Prediction of SMILES failed: {smiles}",
        )

@app.post("/batch-predict", tags=["predictions"])
async def batch_predict(smiles: str = Form(...)):
    """Predict property from list of smiles
    """
    try:
        smiles = smiles.splitlines() # Split text from textarea input
        smiles = [ s.strip() for s in smiles ] # Strip whitespaces
        smiles = [ s for s in smiles if s ] # Remove empty lines

        start_time = perf_counter()
        smiles, prediction, uncertainty = run_prediction(smiles)
        # smiles, prediction, uncertainty = await run_prediction(smiles_list)
        end_time = perf_counter()
        execution_time = end_time - start_time

        result = { "execution_time": execution_time, "smiles": smiles, "pred": prediction, "uncertainty": uncertainty }
        return result
    except:
        raise HTTPException(
            status_code=404,
            detail=f"Batch prediction of SMILES failed: {smiles}",
        )
````

Many thanks!

**Comment by kevingreenman** [comment]
(2) is a known issue (#220), and as @davidegraff said, we would welcome a PR addressing this!
(4) Chemprop should allow you to use the test set for calibration, but this would be a poor machine learning practice because it would allow leakage of test-set information into the uncertainty calibration process. It is recommended that you split your validation set into a standard validation set and a calibration set.

**Comment by kevingreenman** [comment]
(1) could you post a traceback of the error you get?
(3) no, there is no cross validation done during calibration

**Comment by schinto** [comment]
(1) when the arguments include `--calibration_path`  
```python   
 arguments = [
        '--test_path', '/dev/null',
        '--preds_path', '/dev/null',
        '--checkpoint_dir', BASE_DIR.joinpath("model_checkpoint").as_posix(),
        '--features_generator', 'rdkit_2d_normalized',
        '--no_features_scaling',
        '--uncertainty_method', 'ensemble',
        '--calibration_method', 'zelikman_interval',
        '--calibration_path', 'data/calibration_data.csv', # <--- This line
        '--regression_calibrator_metric', 'stdev'
    ]
```
the program stops with the message:
`Data file did not contain all provided smiles columns: [None]. Data file field names are: ['SMILES', 'property']`

(2) Issue (https://github.com/chemprop/chemprop/issues/220) is open since 2021 ?
(4) The first part of my question was not answered: Is the external calibration data set required?
What is not clear to me is that calibration is done during prediction using `make_predictions` and not after training, so that the calibration is saved with the model.

**Comment by schinto** [comment]
Sorry to ask again.
In [this repository](https://github.com/ritvikmath/YouTubeVideoCode/blob/main/Probability%20Calibration.ipynb) the overall calibration process is outlined.
![image](https://user-images.githubusercontent.com/8945782/196454707-8469acea-cc26-4b8f-9a67-baa124b6ed7e.png)
For new predictions
![image](https://user-images.githubusercontent.com/8945782/196454823-903f5183-f7fc-49cc-9ee5-44132f97cf43.png)

How can the calibrated model be used in ChemProp to make predictions for new samples?
Is there a Python code example?

**Comment by kevingreenman** [comment]
(1) I understand your issue now. The behavior you want isn't implemented yet, but this is a good idea. We'll add this to our list of enhancements that we're hoping to add, but we also welcome a PR if you're able to contribute.

(2) Correct. The Chemprop dev team is a group of students and postdocs who are doing our best to maintain and improve the package, but we have limited time, and this issue hasn't been a priority for us or our funders so far.

(4) I apologize for not fully understanding your question at first. Yes, the calibration set is required at prediction time because the calibration is not saved with the model.

**Comment by kevingreenman** [comment]
We are closing all enhancement requests for chemprop v1.x, as we will not have time to implement these before the upcoming release of [chemprop v2.0](https://github.com/chemprop/chemprop/issues/517). If there is still interest in this enhancement, please open a new issue after the release of v2.0.
